packages/editor/src/pages/Editor/components/Main/Header/BrowserAuthDialog.tsx

```diff
@@ -0,0 +1,179 @@
+  state: IState = {};
+
+  componentDidMount() {
+    hideSplashScreen();
+  }
+
+  render() {
+    return (
+      <Dialog
+        hidden={!this.props.isOpen}
+        onDismiss={this.props.cancel}
+        dialogContentProps={{
+          type: DialogType.normal,
+          title: 'Action required',
+        }}
+        modalProps={{
+          isBlocking: true,
+          layerProps: {
+            onLayerDidMount: this.onDialogShown,
+          },
+        }}
+      >
+        {!this.state.authUrl ? (
+          <Label>
+            Please wait while we prepare the authentication dialog. This may take a few
+            seconds...
+          </Label>
+        ) : (
+          <>
+            <Label>
+              To log in with GitHub, please open the following URL in a browser window:
+            </Label>
+            <TextboxClipboardWrapper text={this.state.authUrl} />
+            <br />
+            <Label>
+              After completing the authentication flow, paste in the resulting token:
+            </Label>
+            <TextField
+              placeholder="<Paste token here>"
+              onChange={this.onTokenInput}
+              errorMessage={this.state.errorMessage}
+              iconProps={
+                this.state.decodedToken
+                  ? {
+                      iconName: 'Checkmark',
+                    }
+                  : {}
+              }
+              required={
+                /* show required asterisk until it's already been fulfilled */
+                !this.state.decodedToken
+              }
+            />
+          </>
+        )}
+        <DialogFooter>
+          <PrimaryButton onClick={this.onOk} disabled={!this.shouldAllowOk()} text="OK" />
+          <DefaultButton onClick={this.props.cancel} text="Cancel" />
+        </DialogFooter>
+      </Dialog>
+    );
+  }
+
+  onDialogShown = () => {
+    if (!this.keyGenerationInProgress) {
+      this.keyGenerationInProgress = true;
+
+      forge.pki.rsa.generateKeyPair(
+        {
+          bits: 2048,
+          workers: 2 /* number of web workers to use */,
+        },
+        (err, keypair) => {
+          if (err) {
+            this.setState({ errorMessage: err.toString() });
+            return;
+          }
+
+          const publicKeyString = forge.pki.publicKeyToPem(keypair.publicKey);
+          const privateKeyString = forge.pki.privateKeyToPem(keypair.privateKey);
+
+          this.privateKey = new NodeRSA(privateKeyString);
+
+          this.setState({
+            authUrl:
+              currentEditorUrl +
+              '/#/auth?key=' +
+              encodeURIComponent(btoa(publicKeyString)),
+          });
+        },
+      );
+    }
+
+    // Clear out any previous error state, or the token input
+    this.setState({
+      errorMessage: null,
+      encodedToken: null,
+      decodedToken: null,
+    });
+  };
+
+  shouldAllowOk = () => Boolean(!this.state.errorMessage && this.state.encodedToken);
```

Comment:
you don't like `!!`?

Reply:
Depends.  I'm OK with using `!!` when it's a simple case (esp. for something in React, like `<Only when={!!additionalInfo}>`), but for a complex statement that already has an `!` and a `&&`, I think it's cleaner to do `Boolean`